File: Execution.Tests/ExecutionTests.cs
using Terminal.Execution;

namespace Execution.Tests;

[TestClass]
public class ExecutionTests
{
    private string _root = string.Empty;
    private TestHarness _harness = null!;

    [TestInitialize]
    public void Setup()
    {
        _root = Path.Combine(Path.GetTempPath(), "clr-exec-" + Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(_root);
        Directory.CreateDirectory(Path.Combine(_root, "documents"));
        File.WriteAllText(Path.Combine(_root, "notes.txt"), "hello");

        _harness = new TestHarness(_root);
    }

    [TestCleanup]
    public void Cleanup()
    {
        if (Directory.Exists(_root))
        {
            Directory.Delete(_root, recursive: true);
        }
    }

    // ---- argument binding -------------------------------------------------------
    // The original converter recognised only StringConstant, so every unquoted
    // argument threw NotImplementedException and was then silently swallowed.

    [TestMethod]
    public void UnquotedArgumentBinds()
    {
        var result = _harness.Run("cd documents");

        Assert.IsInstanceOfType(result, typeof(PathValue));
        StringAssert.EndsWith(_harness.PathModule.CurrentPath.TrimEnd(Path.DirectorySeparatorChar), "documents");
    }

    [TestMethod]
    public void QuotedArgumentBinds()
    {
        _harness.Run(@"cd ""documents""");

        StringAssert.EndsWith(_harness.PathModule.CurrentPath.TrimEnd(Path.DirectorySeparatorChar), "documents");
    }

    [TestMethod]
    public void NumericArgumentBindsAsNumber()
    {
        var result = _harness.Run("echo 42");

        Assert.IsInstanceOfType(result, typeof(NumberValue));
        Assert.AreEqual(42d, ((NumberValue)result).Number);
    }

    [TestMethod]
    public void MissingRequiredArgumentIsReported()
    {
        string message = _harness.RunExpectingError("cd");

        StringAssert.Contains(message, "needs an argument");
        StringAssert.Contains(message, "TargetPath");
    }

    [TestMethod]
    public void TooManyArgumentsAreReported()
    {
        string message = _harness.RunExpectingError("cd documents extra");

        // Counts what was supplied rather than what was left over: the old wording read
        // "takes 1 argument(s), but 1 more were given" for a two-argument call.
        StringAssert.Contains(message, "takes 1 argument, but 2 were given");
    }

    [TestMethod]
    public void UnknownNamedArgumentIsReported()
    {
        string message = _harness.RunExpectingError("ls -nonsense value");

        StringAssert.Contains(message, "no argument named");
    }

    [TestMethod]
    public void NamedArgumentBindsOutOfPosition()
    {
        var result = _harness.Run("ls -path documents");

        Assert.IsInstanceOfType(result, typeof(ListValue));
    }

    // ---- pipes ------------------------------------------------------------------
    // ExecuteNominal used to run OrderedCommands.First() and discard the rest.

    [TestMethod]
    public void PipeRunsEveryStage()
    {
        var result = _harness.Run("mkdir alpha | echo");

        // echo receives mkdir's result rather than re-running mkdir.
        Assert.IsInstanceOfType(result, typeof(PathValue));
        Assert.IsTrue(Directory.Exists(Path.Combine(_root, "alpha")));
        StringAssert.EndsWith(((PathValue)result).Path.TrimEnd(Path.DirectorySeparatorChar), "alpha");
    }

    [TestMethod]
    public void PipeThreadsValueBetweenCommands()
    {
        var result = _harness.Run("echo documents | cd");

        Assert.IsInstanceOfType(result, typeof(PathValue));
        StringAssert.EndsWith(_harness.PathModule.CurrentPath.TrimEnd(Path.DirectorySeparatorChar), "documents");
    }

    [TestMethod]
    public void LaterPipeStageFailureIsReported()
    {
        string message = _harness.RunExpectingError("echo nowhere | cd");

        StringAssert.Contains(message, "Directory does not exist");
    }

    // ---- command forms ----------------------------------------------------------
    // Only the CLI form executed before; the other two threw NotImplementedException.

    [TestMethod]
    public void FunctionFormExecutes()
    {
        var result = _harness.Run("echo(hello)");

        Assert.AreEqual("hello", result.ToDisplayString());
    }

    [TestMethod]
    public void FunctionFormWithNamedArgumentExecutes()
    {
        var result = _harness.Run("echo(text: hello)");

        Assert.AreEqual("hello", result.ToDisplayString());
    }

    [TestMethod]
    public void ObjectInstanceFormEvaluates()
    {
        var result = _harness.Run("<measurement unit=metres/>");

        var instance = (ObjectValue)result;
        Assert.AreEqual("measurement", instance.TypeName);
        Assert.AreEqual("metres", instance.Attributes["unit"].ToDisplayString());
    }

    [TestMethod]
    public void NestedObjectInstanceEvaluates()
    {
        var result = _harness.Run("<outer><inner depth=2/></outer>");

        var instance = (ObjectValue)result;
        Assert.AreEqual(1, instance.Children.Count);
        Assert.AreEqual("inner", instance.Children[0].TypeName);
        Assert.AreEqual(2d, ((NumberValue)instance.Children[0].Attributes["depth"]).Number);
    }

    // ---- variables --------------------------------------------------------------
    // Variable had get-only properties and no constructor, so nothing could be bound.

    [TestMethod]
    public void ObjectInstanceBindsVariable()
    {
        _harness.Run("<size|measurement unit=metres/>");

        var variable = _harness.Scope.GetVariable("size");
        Assert.IsNotNull(variable);
        Assert.AreEqual("measurement", ((ObjectValue)variable!.Value).TypeName);
    }

    [TestMethod]
    public void VariableReferenceResolvesAsArgument()
    {
        _harness.Run("<target|folder name=documents/>");
        var result = _harness.Run("echo $target");

        Assert.IsInstanceOfType(result, typeof(ObjectValue));
    }

    [TestMethod]
    public void UnknownVariableIsReported()
    {
        string message = _harness.RunExpectingError("echo $missing");

        StringAssert.Contains(message, "Unknown variable");
    }

    // ---- tags as arguments --------------------------------------------------------
    // The grammar admits a tag wherever an argument is expected. Only a tag standing
    // alone as a pipeline stage used to evaluate; everywhere else reported
    // "Unsupported argument value : TagValue".

    [TestMethod]
    public void TagIsAcceptedAsAnArgument()
    {
        var result = (ObjectValue)_harness.Run("echo <measurement unit=metres/>");

        Assert.AreEqual("measurement", result.TypeName);
        Assert.AreEqual("metres", result.Attributes["unit"].ToDisplayString());
    }

    [TestMethod]
    public void ComponentTagIsAcceptedAsAnArgument() =>
        Assert.IsInstanceOfType(_harness.Run("echo {renderer/}"), typeof(ComponentValue));

    [TestMethod]
    public void TagAsAnArgumentStillBindsItsVariable()
    {
        _harness.Run("echo <size|measurement unit=metres/>");

        Assert.IsInstanceOfType(_harness.Scope.GetVariable("size")!.Value, typeof(ObjectValue));
    }

    [TestMethod]
    public void TagIsAcceptedAsAFunctionArgument() =>
        Assert.IsInstanceOfType(_harness.Run("echo(<thing/>)"), typeof(ObjectValue));

    [TestMethod]
    public void VariableTagIsAcceptedAsAnArgument()
    {
        _harness.Run("<size|measurement unit=metres/>");

        Assert.AreEqual("measurement", ((ObjectValue)_harness.Run("echo <$size>")).TypeName);
    }

    // ---- errors -----------------------------------------------------------------

    [TestMethod]
    public void UnknownCommandIsReported()
    {
        string message = _harness.RunExpectingError("nosuchcommand");

        StringAssert.Contains(message, "Unknown command");
    }

    [TestMethod]
    public void FailingCommandReportsWhyRatherThanSilentlyDoingNothing()
    {
        string message = _harness.RunExpectingError("cd nowhere");

        StringAssert.Contains(message, "Directory does not exist");
    }

    // ---- results ----------------------------------------------------------------

    [TestMethod]
    public void ListDirectoryReturnsEntries()
    {
        var result = (ListValue)_harness.Run("ls");

        var paths = result.Items.Cast<PathValue>().ToList();
        Assert.IsTrue(paths.Any(p => p.Kind == PathKind.Directory && p.Name == "documents"));
        Assert.IsTrue(paths.Any(p => p.Kind == PathKind.File && p.Name == "notes.txt"));
    }

    [TestMethod]
    public void EmptyCommandDoesNothing()
    {
        var result = _harness.Run("");

        Assert.AreSame(RuntimeValue.Empty, result);
    }

    // ---- component tags ----------------------------------------------------------
    // The parser builds these now, so the evaluator has to produce a value for them.

    [TestMethod]
    public void ComponentTagEvaluates()
    {
        var component = (ComponentValue)_harness.Run("{renderer colour=red/}");

        Assert.AreEqual("renderer", component.TypeName);
        Assert.AreEqual("red", component.Attributes["colour"].ToDisplayString());
    }

    [TestMethod]
    public void ComponentTagBindsAVariable()
    {
        _harness.Run("{handle|renderer/}");

        Assert.IsInstanceOfType(_harness.Scope.GetVariable("handle")!.Value, typeof(ComponentValue));
    }

    [TestMethod]
    public void ComponentReadsBackTheWayItWasWritten() =>
        Assert.AreEqual("{renderer colour=red/}", _harness.Run("{renderer colour=red/}").ToDisplayString());

    [TestMethod]
    public void EntityMayContainAComponent()
    {
        var entity = (ObjectValue)_harness.Run("<player>{renderer/}</player>");

        Assert.AreEqual("player", entity.TypeName);
    }

    [TestMethod]
    public void VariableTagReadsTheVariableBack()
    {
        _harness.Run("<size|measurement unit=metres/>");
        var value = (ObjectValue)_harness.Run("<$size>");

        Assert.AreEqual("measurement", value.TypeName);
    }

    [TestMethod]
    public void VariableTagForAnUnknownNameIsReported() =>
        StringAssert.Contains(_harness.RunExpectingError("<$missing>"), "Unknown variable");

    // ---- undo -------------------------------------------------------------------

    [TestMethod]
    public void UndoReversesDirectoryCreation()
    {
        _harness.Run("mkdir beta");
        Assert.IsTrue(Directory.Exists(Path.Combine(_root, "beta")));

        _harness.History.UndoLast();

        Assert.IsFalse(Directory.Exists(Path.Combine(_root, "beta")));
    }

    [TestMethod]
    public void UndoReversesDirectoryChange()
    {
        string before = _harness.PathModule.CurrentPath;
        _harness.Run("cd documents");

        _harness.History.UndoLast();

        Assert.AreEqual(before, _harness.PathModule.CurrentPath);
    }

    [TestMethod]
    public void UndoClearsTheCommandsOutput()
    {
        _harness.Run("mkdir gamma");

        _harness.History.UndoLast();

        Assert.IsTrue(_harness.Output.Cleared);
    }

    [TestMethod]
    public void UndoOnEmptyHistoryIsHarmless()
    {
        _harness.History.UndoLast();

        Assert.AreEqual(0, _harness.History.Count);
    }
}

File: Core.Tests/ExecutionTests.fs
namespace CommandLineReimagined.Core.Tests

open Microsoft.VisualStudio.TestTools.UnitTesting
open CommandLineReimagined.Core
open CommandLineReimagined.Core.Tests.Harness
open CommandLineReimagined.Core.Tests.SessionHarness

/// <summary>Binding, pipes, command forms, tags, variables and errors.</summary>
/// <remarks>
/// The port of `Execution.Tests/ExecutionTests.cs`. Every case is the same case; what
/// differs is what it asserts against. The originals created a temporary directory and
/// checked the disk afterwards, which meant the tests could only say that something
/// happened somewhere. These read the projection, so they say exactly what the line
/// changed, and they run without a filesystem at all.
/// </remarks>
[<TestClass>]
type ExecutionTests() =

    // ------------------------------------------------------------ argument binding

    [<TestMethod>]
    member _.UnquotedArgumentBinds() =
        let harness = seeded ()
        harness.Run "in documents" |> ignore

        Assert.AreEqual<string>("/documents", harness.Location)

    [<TestMethod>]
    member _.QuotedArgumentBinds() =
        let harness = seeded ()
        harness.Run "in \"documents\"" |> ignore

        Assert.AreEqual<string>("/documents", harness.Location)

    [<TestMethod>]
    member _.NumericArgumentBindsAsNumber() =
        let harness = seeded ()

        match harness.Run "echo 42" with
        | Value.Number n -> Assert.AreEqual<float>(42.0, n)
        | other -> Assert.Fail(sprintf "Expected a number, got %A" other)

    [<TestMethod>]
    member _.MissingRequiredArgumentIsReported() =
        let harness = seeded ()
        let message = harness.Error "in"

        StringAssert.Contains(message, "needs an argument")
        StringAssert.Contains(message, "TargetPath")

    /// The count names what was supplied rather than what remains, which used to read
    /// as "takes 1 argument(s), but 1 more were given".
    [<TestMethod>]
    member _.TooManyArgumentsAreReported() =
        let harness = seeded ()

        StringAssert.Contains(harness.Error "in documents extra", "takes 1 argument, but 2 were given")

    [<TestMethod>]
    member _.UnknownNamedArgumentIsReported() =
        let harness = seeded ()

        StringAssert.Contains(harness.Error "ls -nonsense value", "no argument named")

    [<TestMethod>]
    member _.NamedArgumentBindsOutOfPosition() =
        let harness = seeded ()

        Assert.AreEqual<string>("notes.txt", harness.Names "ls -path documents")

    [<TestMethod>]
    member _.AFaultCarriesItsKind() =
        let harness = seeded ()

        Assert.AreEqual<FaultKind>(Binding, (harness.Fail "in").Kind)
        Assert.AreEqual<FaultKind>(NotFound, (harness.Fail "in nowhere").Kind)
        Assert.AreEqual<FaultKind>(UnknownCommand, (harness.Fail "nosuchcommand").Kind)

    /// The stage number is stamped by the evaluator, which is the only thing that
    /// knows where in the line the failure was.
    [<TestMethod>]
    member _.AFaultKnowsWhichStageFailed() =
        let harness = seeded ()

        Assert.AreEqual<int option>(Some 1, (harness.Fail "in nowhere").Stage)
        Assert.AreEqual<int option>(Some 2, (harness.Fail "echo nowhere | in").Stage)

    // ------------------------------------------------------------------------ pipes

    [<TestMethod>]
    member _.PipeRunsEveryStage() =
        let harness = seeded ()

        // echo receives mkdir's result rather than re-running mkdir.
        match harness.Run "mkdir alpha | echo" with
        | Value.File file -> Assert.AreEqual<string>("alpha", file.Name)
        | other -> Assert.Fail(sprintf "Expected a file, got %A" other)

        Assert.IsTrue(harness.Exists "alpha")

    [<TestMethod>]
    member _.PipeThreadsValueBetweenCommands() =
        let harness = seeded ()
        harness.Run "echo documents | in" |> ignore

        Assert.AreEqual<string>("/documents", harness.Location)

    [<TestMethod>]
    member _.LaterPipeStageFailureIsReported() =
        let harness = seeded ()

        StringAssert.Contains(harness.Error "echo nowhere | in", "Directory does not exist")

    /// <summary>A line is one transaction (decision 0015).</summary>
    /// <remarks>
    /// The case the record was written for. The `mkdir` succeeded and the `in` did not,
    /// and the folder must not survive: a failed line leaves no trace, so the user does
    /// not have to know how many stages ran before the failure in order to undo it.
    /// </remarks>
    [<TestMethod>]
    member _.AFailedLineLeavesNothingBehind() =
        let harness = seeded ()

        StringAssert.Contains(harness.Error "mkdir a | in nowhere", "Directory does not exist")
        Assert.IsFalse(harness.Exists "a", "The mkdir should have been rolled back with the line.")
        Assert.AreEqual<string list>([ "seed" ], harness.History())

    /// A later stage sees what an earlier one did, because the stages share a working
    /// projection even though nothing has been committed yet.
    [<TestMethod>]
    member _.ALaterStageSeesAnEarlierStagesEffect() =
        let harness = seeded ()
        harness.Run "mkdir scratch | in" |> ignore

        Assert.AreEqual<string>("/scratch", harness.Location)

    [<TestMethod>]
    member _.AWholeLineIsUndoneAtOnce() =
        let harness = seeded ()
        harness.Run "mkdir one | in" |> ignore

        harness.Run "undo" |> ignore

        Assert.IsFalse(harness.Exists "/one")
        Assert.AreEqual<string>("/", harness.Location)

    // --------------------------------------------------------------- command forms

    [<TestMethod>]
    member _.FunctionFormExecutes() =
        let harness = seeded ()

        Assert.AreEqual<string>("hello", harness.Text "echo(hello)")

    [<TestMethod>]
    member _.FunctionFormWithNamedArgumentExecutes() =
        let harness = seeded ()

        Assert.AreEqual<string>("hello", harness.Text "echo(text: hello)")

    [<TestMethod>]
    member _.ObjectInstanceFormEvaluates() =
        let harness = seeded ()

        match harness.Run "<measurement unit=metres/>" with
        | Value.Object tag ->
            Assert.AreEqual<string>("measurement", tag.TypeName)
            Assert.AreEqual<string>("metres", Value.display tag.Attributes["unit"])
        | other -> Assert.Fail(sprintf "Expected an object, got %A" other)

    [<TestMethod>]
    member _.NestedObjectInstanceEvaluates() =
        let harness = seeded ()

        match harness.Run "<outer><inner depth=2/></outer>" with
        | Value.Object tag ->
            Assert.AreEqual<int>(1, tag.Children.Length)

            match tag.Children[0] with
            | Value.Object inner ->
                Assert.AreEqual<string>("inner", inner.TypeName)
                Assert.AreEqual<Value>(Value.Number 2.0, inner.Attributes["depth"])
            | other -> Assert.Fail(sprintf "Expected a nested object, got %A" other)
        | other -> Assert.Fail(sprintf "Expected an object, got %A" other)

    [<TestMethod>]
    member _.EmptyCommandDoesNothing() =
        let harness = seeded ()

        Assert.AreEqual<Value>(Value.Empty, harness.Run "")

    // ------------------------------------------------------------------- variables

    /// <summary>A tag that names a variable binds it, as an event.</summary>
    /// <remarks>
    /// The old implementation mutated the scope where it stood. Now it is a
    /// `VariableChanged` in the line's transaction, so the binding survives a reload
    /// and comes back out on `undo` with everything else the line did.
    /// </remarks>
    [<TestMethod>]
    member _.ObjectInstanceBindsVariable() =
        let harness = seeded ()
        harness.Run "<size|measurement unit=metres/>" |> ignore

        match harness.Variable "size" with
        | Some(Value.Object tag) -> Assert.AreEqual<string>("measurement", tag.TypeName)
        | other -> Assert.Fail(sprintf "Expected $size to be an object, got %A" other)

    [<TestMethod>]
    member _.UndoOfATagBindingUnbindsTheVariable() =
        let harness = seeded ()
        harness.Run "<size|measurement unit=metres/>" |> ignore

        harness.Run "undo" |> ignore

        Assert.IsTrue((harness.Variable "size").IsNone)

    [<TestMethod>]
    member _.VariableReferenceResolvesAsArgument() =
        let harness = seeded ()
        harness.Run "<target|folder name=documents/>" |> ignore

        match harness.Run "echo $target" with
        | Value.Object _ -> ()
        | other -> Assert.Fail(sprintf "Expected an object, got %A" other)

    [<TestMethod>]
    member _.UnknownVariableIsReported() =
        let harness = seeded ()

        StringAssert.Contains(harness.Error "echo $missing", "Unknown variable")

    // -------------------------------------------------------- tags as arguments

    [<TestMethod>]
    member _.TagIsAcceptedAsAnArgument() =
        let harness = seeded ()

        match harness.Run "echo <measurement unit=metres/>" with
        | Value.Object tag ->
            Assert.AreEqual<string>("measurement", tag.TypeName)
            Assert.AreEqual<string>("metres", Value.display tag.Attributes["unit"])
        | other -> Assert.Fail(sprintf "Expected an object, got %A" other)

    [<TestMethod>]
    member _.ComponentTagIsAcceptedAsAnArgument() =
        let harness = seeded ()

        match harness.Run "echo {renderer/}" with
        | Value.Component _ -> ()
        | other -> Assert.Fail(sprintf "Expected a component, got %A" other)

    [<TestMethod>]
    member _.TagAsAnArgumentStillBindsItsVariable() =
        let harness = seeded ()
        harness.Run "echo <size|measurement unit=metres/>" |> ignore

        match harness.Variable "size" with
        | Some(Value.Object _) -> ()
        | other -> Assert.Fail(sprintf "Expected $size to be an object, got %A" other)

    [<TestMethod>]
    member _.TagIsAcceptedAsAFunctionArgument() =
        let harness = seeded ()

        match harness.Run "echo(<thing/>)" with
        | Value.Object _ -> ()
        | other -> Assert.Fail(sprintf "Expected an object, got %A" other)

    [<TestMethod>]
    member _.VariableTagIsAcceptedAsAnArgument() =
        let harness = seeded ()
        harness.Run "<size|measurement unit=metres/>" |> ignore

        match harness.Run "echo <$size>" with
        | Value.Object tag -> Assert.AreEqual<string>("measurement", tag.TypeName)
        | other -> Assert.Fail(sprintf "Expected an object, got %A" other)

    // ------------------------------------------------------------- component tags

    [<TestMethod>]
    member _.ComponentTagEvaluates() =
        let harness = seeded ()

        match harness.Run "{renderer colour=red/}" with
        | Value.Component tag ->
            Assert.AreEqual<string>("renderer", tag.TypeName)
            Assert.AreEqual<string>("red", Value.display tag.Attributes["colour"])
        | other -> Assert.Fail(sprintf "Expected a component, got %A" other)

    [<TestMethod>]
    member _.ComponentTagBindsAVariable() =
        let harness = seeded ()
        harness.Run "{handle|renderer/}" |> ignore

        match harness.Variable "handle" with
        | Some(Value.Component _) -> ()
        | other -> Assert.Fail(sprintf "Expected $handle to be a component, got %A" other)

    [<TestMethod>]
    member _.ComponentReadsBackTheWayItWasWritten() =
        let harness = seeded ()

        Assert.AreEqual<string>("{renderer colour=red/}", harness.Text "{renderer colour=red/}")

    [<TestMethod>]
    member _.EntityMayContainAComponent() =
        let harness = seeded ()

        match harness.Run "<player>{renderer/}</player>" with
        | Value.Object tag -> Assert.AreEqual<string>("player", tag.TypeName)
        | other -> Assert.Fail(sprintf "Expected an object, got %A" other)

    [<TestMethod>]
    member _.VariableTagReadsTheVariableBack() =
        let harness = seeded ()
        harness.Run "<size|measurement unit=metres/>" |> ignore

        match harness.Run "<$size>" with
        | Value.Object tag -> Assert.AreEqual<string>("measurement", tag.TypeName)
        | other -> Assert.Fail(sprintf "Expected an object, got %A" other)

    [<TestMethod>]
    member _.VariableTagForAnUnknownNameIsReported() =
        let harness = seeded ()

        StringAssert.Contains(harness.Error "<$missing>", "Unknown variable")

    // ---------------------------------------------------------------------- errors

    [<TestMethod>]
    member _.UnknownCommandIsReported() =
        let harness = seeded ()

        StringAssert.Contains(harness.Error "nosuchcommand", "Unknown command")

    /// The command that reports unknown names is not itself a name you can type: typed,
    /// it reported an unknown command with no name.
    [<TestMethod>]
    member _.TypingTheUnknownCommandsNameNamesIt() =
        let harness = seeded ()

        let fault = harness.Fail "UnknownCommand"

        Assert.AreEqual<FaultKind>(UnknownCommand, fault.Kind)
        Assert.AreEqual<string>("Unknown command : UnknownCommand", fault.Message)

    [<TestMethod>]
    member _.FailingCommandReportsWhyRatherThanSilentlyDoingNothing() =
        let harness = seeded ()

        StringAssert.Contains(harness.Error "in nowhere", "Directory does not exist")

    /// A parse failure is a fault like any other, so the session boundary never raises.
    [<TestMethod>]
    member _.ASyntaxErrorIsAFaultRatherThanAnException() =
        let harness = seeded ()

        Assert.AreEqual<FaultKind>(Syntax, (harness.Fail "<thing").Kind)

    // --------------------------------------------------------------------- results

    [<TestMethod>]
    member _.ListDirectoryReturnsEntries() =
        let harness = seeded ()

        let names = harness.Column "ls" "name"

        assertContains "documents" names
        assertContains "readme.txt" names

    /// <summary>Folders first, then files, each by name (decision 0016).</summary>
    /// <remarks>
    /// Normative now. The old specification said the order "follows the filesystem",
    /// which meant it was whatever the host returned and could not be tested at all.
    /// </remarks>
    [<TestMethod>]
    member _.AListingIsFoldersThenFilesEachByName() =
        let harness = seeded ()
        harness.Run "mkdir zebra" |> ignore
        harness.Run "write apple.txt a" |> ignore

        Assert.AreEqual<string>("documents examples guide projects zebra apple.txt readme.txt", harness.Names "ls")

    /// <summary>A listing is records and nothing else (Phase 3).</summary>
    /// <remarks>
    /// Until Phase 3 a listing below the root began with a parent entry that navigated
    /// rather than naming a record. A table has nowhere to put one — every row is a
    /// record — so the page offers `out` in the location line instead, and the `out`
    /// command is what it runs.
    /// </remarks>
    [<TestMethod>]
    member _.AListingHasNoParentRow() =
        let harness = seeded ()

        Assert.AreEqual<string>("documents examples guide projects readme.txt", harness.Names "ls")

        harness.Run "in documents" |> ignore
        Assert.AreEqual<string>("notes.txt", harness.Names "ls")

    /// `ls | in` with no quoting rule: the parent entry argues where it goes.
    [<TestMethod>]
    member _.TheParentEntryNavigates() =
        let harness = seeded ()
        harness.Run "in documents" |> ignore

        harness.Run "out" |> ignore

        Assert.AreEqual<string>("/", harness.Location)

    // ------------------------------------------------------------ value stages

    /// Decision 0032: a variable standing alone is a line, and its value is the answer.
    [<TestMethod>]
    member _.AVariableStandingAloneAnswersItsValue() =
        let harness = seeded ()
        harness.Run "set v 5" |> ignore

        Assert.AreEqual<Value>(Value.Number 5.0, harness.Run "$v")
        Assert.AreEqual<string>("5", harness.Text "$v")

    /// A value stage is the head of a pipeline like any other, so what it holds flows on.
    [<TestMethod>]
    member _.AVariableStageFeedsThePipe() =
        let harness = seeded ()
        harness.Run "ls | set files" |> ignore

        Assert.AreEqual<Value>(Value.Number 5.0, harness.Run "$files | count")
        Assert.AreEqual<string>("documents examples guide projects", harness.Names "$files | where $row.kind eq folder")

    /// Members read off a value stage the way they read off an argument.
    [<TestMethod>]
    member _.AVariableStageReadsItsMembers() =
        let harness = seeded ()
        harness.Run "try read missing.txt | set problem" |> ignore

        Assert.AreEqual<string>("NotFound", harness.Text "$problem.kind")

    /// A value stage changes nothing, so a live view may re-read it.
    [<TestMethod>]
    member _.AVariableStageOnlyReads() =
        let harness = seeded ()
        harness.Run "set v 5" |> ignore

        let response = harness.Refresh "$v"

        Assert.IsTrue(response.Fault.IsNone)
        Assert.AreEqual<Value option>(Some(Value.Number 5.0), response.Result)

    [<TestMethod>]
    member _.AnUnknownVariableStageIsReported() =
        let harness = seeded ()

        Assert.AreEqual<string>("Unknown variable: $nope", harness.Error "$nope")
        Assert.AreEqual<string>("Unknown variable: $nope", harness.Error "echo $nope")

    /// Decision 0032: `$row` outside a predicate says where it does exist.
    [<TestMethod>]
    member _.RowStandingAloneSaysWhereItExists() =
        let harness = seeded ()

        let fault = harness.Fail "$row"

        Assert.AreEqual<string>(
            "$row is the row a predicate is testing. It exists only inside where, find, in and save-view: ls | where $row.kind eq folder.",
            fault.Message)
        Assert.AreEqual(NotFound, fault.Kind)

    /// The same for `$row` read as an argument, which used to be an unknown variable.
    [<TestMethod>]
    member _.RowAsAnArgumentSaysWhereItExists() =
        let harness = seeded ()

        StringAssert.StartsWith(harness.Error "echo $row", "$row is the row a predicate is testing.")
        StringAssert.StartsWith(harness.Error "$row.kind", "$row is the row a predicate is testing.")

    /// Inside a predicate `$row` is the row, as it always was.
    [<TestMethod>]
    member _.RowInsideAPredicateIsTheRow() =
        let harness = seeded ()

        Assert.AreEqual<string>("readme.txt", harness.Names "ls | where $row.kind eq text")
